src/stdune/io.ml

```diff
@@ -79,14 +84,23 @@ struct
       in
       try loop () with End_of_file -> Buffer.contents buffer
     in
-    fun t ->
+    fun ?binary t ->
+      let binary = mode binary in
       (* Optimisation for regular files: if the channel supports seeking, we
          compute the length of the file so that we read exactly what we need and
          avoid an extra memory copy. We expect that most files Dune reads are
          regular files so this optimizations seems worth it. *)
-      match in_channel_length t with
-      | exception _ -> read_all_generic t (Buffer.create chunk_size)
-      | n -> (
+      let seek =
+        match (Stdlib.Sys.os_type, binary) with
+        | ("Cygwin" | "Win32"), false -> None
```

Comment:
Can you please leave a comment explaining why windows needs a special case?

Reply:
I'm assuming it's because the channel length is not in agreement with the actual number of characters read.

I've always been suspicious of this `in_channel_length` heuristic. Maybe we can make it more robust.
For example, what if we use this `eagerly_input_string` instead of `really_input_string`?

```
let rec eagerly_input_acc ic s pos len acc =
  if len <= 0 then acc else begin
    let r = input ic s pos len in
    if r = 0
    then acc
    else eagerly_input_acc ic s (pos + r) (len - r) (acc + r)
  end

let bytes_sub_to_string src pos len =
  Bytes.unsafe_to_string (Bytes.sub src pos len)

let eagerly_input_string ic len =
  let buf = Bytes.create len in
  let r = eagerly_input_acc ic buf 0 len 0 in
  if r = len then Bytes.unsafe_to_string buf else
    bytes_sub_to_string buf 0 r
```

Reply:
> I'm assuming it's because the channel length is not in agreement with the actual number of characters read.

Even worse. Calling `in_channel_length` raises `Eof` https://github.com/rgrinberg/jbuilder/runs/977261847?check_suite_focus=true#step:11:1

Reply:
Indeed - the problem is that `in_channel_length` and `seek` don't (and never will) take `\r\n` translations (i.e. text mode) into account.

Reply:
@rgrinberg /sigh.
Ok, I give up. Special-casing windows seems fine.

Reply:
@rgrinberg, wait a second, your link does not support your claim. The error points to line 90, which is `really_input_string`. Do you agree?

Reply:
Yes, you are right.

Reply:
I moved this to a separate PR: #3708.